strawberryfields/apps/train/cost.py

```diff
@@ -0,0 +1,129 @@
+    .. math::
+
+        \partial_\theta KL(\theta) = - \sum_{k=1}^m(\langle n_k\rangle_{\text{data}-
+        \langle n_k\rangle)_{\text{GBS}}\partial_\theta w_k,
+
+    where :math:`\langle n_k\rangle)` denotes the average photon numbers in mode *k*. This class
+    provides methods to compute gradients and evaluate the cost function.
+
+    **Example usage**
+
+    >>> embedding = embed.Exp(4)
+    >>> A = np.ones((4, 4))
+    >>> vgbs = param.VGBS(A, 3, embedding, threshold=True)
+    >>> params = np.array([0.05, 0.1, 0.02, 0.01])
+    >>> data = np.zeros((4, 4))
+    >>> kl = cost.KL(data, vgbs)
+    >>> kl.evaluate(params)
+    -0.2866830267216749
+    >>> kl.grad(params)
+    array([-0.52812574, -0.5201932 , -0.53282312, -0.53437824])
+
+    Args:
+        data (np.array): Array of samples representing the training data
+        vgbs: Variational GBS class
+
+    """
+
+    def __init__(self, data: np.ndarray, vgbs):
+        self.data = data
+        self.vgbs = vgbs
+        self.nr_samples = len(data)
+        self.nr_modes = len(data[0])
+
+    def mean_n_data(self):
+        r"""Computes the average number of photons :math:`\langle n_k\rangle_{\text{data}` from
+        the data. When the data are samples from threshold detectors, this gives the average number
+        of clicks.
+
+        **Example usage**
+
+        >>> kl.mean_n_data()
+        array([0., 0., 0., 0.])
+
+        Returns:
+            array: vector of mean photon numbers per mode
+        """
+        return np.sum(self.data, axis=0) / self.nr_samples
+
+    def grad(self, params: np.ndarray) -> np.ndarray:
+        """Calculates the gradient of the Kullback-Liebler cost function with respect to the
+        trainable parameters
+
+        **Example usage**
+
+        >>> kl.grad(params)
+        array([-0.52812574, -0.5201932 , -0.53282312, -0.53437824])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+        Returns:
+            array: the gradient of the K-L cost function with respect to :math:`\theta`
+        """
+        weights = self.vgbs.embedding(params)
+        if self.vgbs.threshold:
+            n_diff = self.vgbs.mean_clicks_by_mode(params) - self.mean_n_data()
+        else:
+            n_diff = self.vgbs.mean_photons_by_mode(params) - self.mean_n_data()
+        return (n_diff / weights) @ self.vgbs.embedding.jacobian(params)
+
+    def evaluate(self, params: np.ndarray) -> float:
+        """Computes the value of the Kullback-Liebler divergence cost function.
+
+        **Example usage**
+
+        >>> kl.evaluate(params)
+        -0.2866830267216749
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+        Returns:
+            float: the value of the cost function
+        """
+        kl = 0
+        for sample in self.data:
+            kl += np.log(self.vgbs.prob_sample(params, sample))
+        return -kl / self.nr_samples - np.log(self.nr_samples)
```

Comment:
I think I'm still confused on the equation for KL divergence :thinking:

Reply:
In its general form, the KL divergence reads
`KL = \sum_s Q(s) log[Q(s)/P(s)]`
Now, let `P` be the GBS distribution and `Q` be the data distribution. We don't have access to `Q` in closed form, all we have is a bunch of data, from which we can estimate the data distribution as `Q(s) = t_s/T`, where `t_s` is the number of times the sample `s` appears in the data and `T` is the total number of samples. Then we have
```
KL = \sum_s (t_s/T) log [t_s/ (T P(s)]
= -1/T \sum_s t_s [log[P(s)] + log(T/t_s)]
= -1/T \sum_{s\in data} [log[P(s)] - constant 
```
I need to double check this constant term, but it doesn't matter and should probably be removed anyway. Up to this constant, the expression coincides with the calculation in the code